### DISCUSSION TITLE
Use model struct as method parameter

### DISCUSSION THREAD

**Discussion by maratori** [sqlc-dev/sqlc#875] [Feature Requests & Ideas]
Some of insert queries contains all columns of table as input parameters.
In that case generated method may apply model struct instead of `...Params` struct.

For example following method
https://github.com/kyleconroy/sqlc/blob/bda1a0089e042a33f64f9dd1b59a2c3631fb659e/examples/ondeck/postgresql/city.sql.go#L28
may easily become
```go
func (q *Queries) CreateCity(ctx context.Context, arg City) (City, error) {
```

Because table contains two columns
https://github.com/kyleconroy/sqlc/blob/bda1a0089e042a33f64f9dd1b59a2c3631fb659e/examples/ondeck/postgresql/schema/0001_city.sql#L1-L4
and the query uses all of them
https://github.com/kyleconroy/sqlc/blob/bda1a0089e042a33f64f9dd1b59a2c3631fb659e/examples/ondeck/postgresql/query/city.sql#L15-L21

### Implementation
For output parameters there is struct "deduplication"
https://github.com/kyleconroy/sqlc/blob/bda1a0089e042a33f64f9dd1b59a2c3631fb659e/internal/codegen/golang/result.go#L183

I believe we need to introduce the same deduplication for input parameters
https://github.com/kyleconroy/sqlc/blob/bda1a0089e042a33f64f9dd1b59a2c3631fb659e/internal/codegen/golang/result.go#L166

**Comment by ryansmith3136** [comment]
Hi. Has anything become of this idea?

**Reply by kyleconroy** [reply]
Nothing has been decided. If we do add support for this, it will need to go behind a configuration flag so that we don't break existing applications.

**Reply by Crocmagnon** [reply]
I would love to see this implemented. Sometimes it's also easier to pass the full entity around instead of having to create another for the occasion of executing a DB query, even if the query doesn't use all fields of the entity.

**Reply by omertoast** [reply]
This can make creating queries for seeding the database a lot easier, otherwise, you'll end up with two `Create` queries with duplicated insert fields. This is not a problem if you have few columns but when you have 50 columns in a table, you need to keep track of both queries and update both when you add a new column to the table. 

I'd like to know if there is any way right now to do this nicely.

**Comment by ecsdavidt** [comment]
Seems like a flag in `query.sql`, to note that you'd like this function to persist or update a whole entity rather than using a params struct, would be really useful.  something akin to  
```sql
-- name: PersistNewCity :entity :one
INSERT INTO city ( * ) VALUES ( ** )
RETURNING *;
```
results in 
```go
func (q *Queries) PersistNewCity(ctx context.Context, arg City) (City, error) {
```
where expands `**` (or whatever) into all columns except the primary key.

**Comment by tloriato** [comment]
+1

How do maintainers feel about it? We can probably help open a PR

**Comment by ciricc** [comment]
We really need it
